### PR TITLE
Two @objcImpl protocol fixes

### PR DESCRIPTION
**Pull request by beccadax**
Fixes two issues with `@_objcImplementation` and protocols specified in the header file:

* IRGen was incorrectly walking the implementation context instead of the imported interface to emit conformances.
* TypeCheckDeclObjC was insisting on having an implementation for protocol requirements even when they were marked `@optional` on the ObjC side.

Fixes rdar://108482165 and rdar://108461441.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci please test

File: test/IRGen/objc_implementation.swift [hunk 1]
```diff
@@ -47,6 +49,14 @@
   final var implProperty2: NSObject?
   
   @objc func mainMethod(_: Int32) { print(implProperty) }
+
+  @objc(copyWithZone:)
+  func copy(with zone: NSZone) -> Any {
```
Comment (tshortli) [inline test/IRGen/objc_implementation.swift:54]: Would it make sense to also have an Interpreter test that exercises the emitted Obj-C protocol conformances, in addition to verifying the IR?

**Review by tshortli — APPROVED** [review]
LGTM overall. Do we need to be concerned at all about the interaction between correctly emitting conformance metadata now and the existing workarounds that adopters have used where they add an empty `@implemenation` block in Obj-C in order to get clang to emit conformance metadata? I don't know enough about the metadata to know if the metadata would be uniqued by the linker or if language runtimes would handle multiple copies gracefully.

File: test/IRGen/objc_implementation.swift (hunk 1 above)
Comment (beccadax) [reply test/IRGen/objc_implementation.swift:54]: I think verifying that the metadata is there will be enough.

**Pull request merged**